File: Data Structures & Algorithms/house-robber/submission-6.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1: 
            return nums[0]
        elif len(nums) == 2:
            return max(nums[0], nums[1])
        else:
            dp1, dp2 = 0, 0
            for n in nums:
                new = max(n+dp1, dp2)
                dp2, dp1 = new, dp2
            return dp2

File: Data Structures & Algorithms/maximum-product-subarray/submission-6.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
            int n = nums.size();
            int ans = INT_MIN;
            int preProd = 1;
            int postProd = 1;

            for (int i = 0; i < n; i++) {
                preProd = (preProd==0)? 1:preProd;
                postProd = (postProd==0)? 1:postProd;

                preProd *= nums[i];
                postProd *= nums[n-1-i];
                ans = max({ans, preProd, postProd});
            }
            return ans;
    }
};
